File: benchmarks/java/src/main/java/dev/jsonatapy/bench/JsonCompare.java
package dev.jsonatapy.bench;

import com.fasterxml.jackson.databind.JsonNode;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

/**
 * Semantic JSON equality: numbers compared as double with relative tolerance
 * (JSONata numbers are IEEE doubles; implementations differ on int-vs-double
 * node types), object key order ignored, array order significant.
 */
public final class JsonCompare {
    private JsonCompare() {}

    public static boolean semanticEquals(JsonNode a, JsonNode b) {
        if (a == null || b == null) {
            return a == b;
        }
        if (a.isNumber() && b.isNumber()) {
            double x = a.doubleValue();
            double y = b.doubleValue();
            return Math.abs(x - y) <= 1e-9 * Math.max(1.0, Math.max(Math.abs(x), Math.abs(y)));
        }
        if (a.isArray() && b.isArray()) {
            if (a.size() != b.size()) {
                return false;
            }
            for (int i = 0; i < a.size(); i++) {
                if (!semanticEquals(a.get(i), b.get(i))) {
                    return false;
                }
            }
            return true;
        }
        if (a.isObject() && b.isObject()) {
            if (a.size() != b.size()) {
                return false;
            }
            Set<String> keys = new HashSet<>();
            for (Iterator<String> it = a.fieldNames(); it.hasNext(); ) {
                keys.add(it.next());
            }
            for (String k : keys) {
                if (!b.has(k) || !semanticEquals(a.get(k), b.get(k))) {
                    return false;
                }
            }
            return true;
        }
        return a.equals(b);
    }
}

File: benchmarks/dotnet/JsonCompare.cs
using System.Text.Json;

namespace JsonataFfiBench;

/// <summary>
/// Semantic JSON equality: numbers compared as double with relative
/// tolerance, object key order ignored, array order significant.
/// (Mirror of the Java JsonCompare.)
/// </summary>
public static class JsonCompare
{
    public static bool SemanticEquals(JsonElement a, JsonElement b)
    {
        if (a.ValueKind == JsonValueKind.Number && b.ValueKind == JsonValueKind.Number)
        {
            double x = a.GetDouble();
            double y = b.GetDouble();
            return Math.Abs(x - y) <= 1e-9 * Math.Max(1.0, Math.Max(Math.Abs(x), Math.Abs(y)));
        }
        if (a.ValueKind != b.ValueKind)
        {
            return false;
        }
        switch (a.ValueKind)
        {
            case JsonValueKind.Array:
            {
                if (a.GetArrayLength() != b.GetArrayLength())
                {
                    return false;
                }
                using var ea = a.EnumerateArray().GetEnumerator();
                using var eb = b.EnumerateArray().GetEnumerator();
                while (ea.MoveNext() && eb.MoveNext())
                {
                    if (!SemanticEquals(ea.Current, eb.Current))
                    {
                        return false;
                    }
                }
                return true;
            }
            case JsonValueKind.Object:
            {
                var bProps = new Dictionary<string, JsonElement>();
                foreach (var p in b.EnumerateObject())
                {
                    bProps[p.Name] = p.Value;
                }
                int aCount = 0;
                foreach (var p in a.EnumerateObject())
                {
                    aCount++;
                    if (!bProps.TryGetValue(p.Name, out JsonElement bv) || !SemanticEquals(p.Value, bv))
                    {
                        return false;
                    }
                }
                return aCount == bProps.Count;
            }
            case JsonValueKind.String:
                return a.GetString() == b.GetString();
            default: // True/False/Null
                return true;
        }
    }
}
